Update changelog for v0.99.0 (#4721)
signalfx/splunk-otel-collector: CHANGELOG.md

## Unreleased

## v0.99.0

### 🛑 Breaking changes 🛑

- (Splunk) `receiver/discovery`: Update the component to emit entity events 

>   - The `log_endpoints` config option has been removed. Endpoints are now only reported that match the configured receiver rules, and are now emitted as entity events.

Author: ```suggestion
  - The `log_endpoints` config option has been removed. Endpoints are now only reported if they match the configured receiver rules, and are now emitted as entity events.
```

>     ([#4692](https://github.com/signalfx/splunk-otel-collector/pull/4692)), ([#4684](https://github.com/signalfx/splunk-otel-collector/pull/4684)),
>     ([#4684](https://github.com/signalfx/splunk-otel-collector/pull/4684)), and ([#4691](https://github.com/signalfx/splunk-otel-collector/pull/4691))

Author: ```suggestion
    ([#4692](https://github.com/signalfx/splunk-otel-collector/pull/4692), [#4684](https://github.com/signalfx/splunk-otel-collector/pull/4684),
    [#4684](https://github.com/signalfx/splunk-otel-collector/pull/4684), and [#4691](https://github.com/signalfx/splunk-otel-collector/pull/4691))
```

  - Emit entity events only for matching receivers ([#4691](https://github.com/signalfx/splunk-otel-collector/pull/4691))
  - Remove `log_endpoints` config option ([#4692](https://github.com/signalfx/splunk-otel-collector/pull/4692))
- (Core) `telemetry`: Distributed internal metrics across different levels. ([#7890](https://github.com/open-telemetry/opentelemetry-collector/pull/7890))
  The internal metrics levels are updated along with reported metrics:
  - The default level is changed from `basic` to `normal`, which can be overridden with `service::telmetry::metrics::level` configuration.
  - Batch processor metrics are updated to be reported starting from `normal` level:
    - `processor_batch_batch_send_size`
    - `processor_batch_metadata_cardinality`
    - `processor_batch_timeout_trigger_send`
    - `processor_batch_size_trigger_send`
  - GRPC/HTTP server and client metrics are updated to be reported starting from `detailed` level:
    - http.client.* metrics
    - http.server.* metrics
    - rpc.server.* metrics
    - rpc.client.* metrics

>   - Note: These metrics are all excluded by default in the Splunk distribution of the Opentelemetry Collector.

Author: ```suggestion
  - Note: These metrics are all excluded by default in the Splunk distribution of the OpenTelemetry Collector.
```

## Unreleased

## v0.99.0

### 🛑 Breaking changes 🛑

- (Splunk) `receiver/discovery`: Update the component to emit entity events 
  - Emit entity events for discovered endpoints with log_endpoints: true ([#4684](https://github.com/signalfx/splunk-otel-collector/pull/4684))
  - Ensure active endpoints emitted as entity events periodically ([#4684](https://github.com/signalfx/splunk-otel-collector/pull/4684))
  - Emit entity events only for matching receivers ([#4691](https://github.com/signalfx/splunk-otel-collector/pull/4691))
  - Remove `log_endpoints` config option ([#4692](https://github.com/signalfx/splunk-otel-collector/pull/4692))
- (Core)`builder`: Add strict version checking when using the builder. Add the temporary flag `--skip-strict-versioning `for skipping this check. ([#9896](https://github.com/open-telemetry/opentelemetry-collector/pull/9896))
    Strict version checking will error on major and minor version mismatches
    between the `otelcol_version` configured and the builder version or versions
    in the go.mod. This check can be temporarily disabled by using the `--skip-strict-versioning`

>     flag. This flag will be removed in a future minor version.

Reviewer: We don't ship the builder. I don't think this should be in our changelog

- (Core) `telemetry`: Distributed internal metrics across different levels. ([#7890](https://github.com/open-telemetry/opentelemetry-collector/pull/7890))
  The internal metrics levels are updated along with reported metrics:
  - The default level is changed from `basic` to `normal`, which can be overridden with `service::telmetry::metrics::level` configuration.
  - Batch processor metrics are updated to be reported starting from `normal` level:
    - `processor_batch_batch_send_size`
    - `processor_batch_metadata_cardinality`
    - `processor_batch_timeout_trigger_send`
    - `processor_batch_size_trigger_send`
  - GRPC/HTTP server and client metrics are updated to be reported starting from `detailed` level:
    - http.client.* metrics
    - http.server.* metrics
    - rpc.server.* metrics

>     - rpc.client.* metrics

Reviewer: We exclude all these metrics in our distro by default. We need to mention that here I believe:

```
All these metrics were already excluded in our distro by default. So this change only affects users who reconfigured the default dropping rules (`metric_relabel_configs`) in the prometheus receiver scraping internal metrics.
```

- (Contrib) `extension/filestorage`: Replace path-unsafe characters in component names ([#3148](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/3148))
  The feature gate `extension.filestorage.replaceUnsafeCharacters` is now stable and cannot be disabled.
  See the File Storage extension's README for details.

> - (Contrib) `all`: Bump minimum version to go 1.21.0 ([#32451](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/32451))

Reviewer: Not, a user-facing change. I don't think we should include go API changes here

## Unreleased

## v0.99.0

### 🛑 Breaking changes 🛑

> - (Splunk) `receiver/discovery`: Update the component to emit entity events 

Author: I'd like some more clarity on the `receiver/discovery` breaking changes as far as what's required (if anything) from users.

Reviewer 2: @dmitryax would know most about this. From the listed PRs, the `log_endpoints` option removal seems user-facing but what are the chances that users have customized receiver/discovery setup?

Author: From Dmitrii's comment, nothing is required of users, but the changelog comment has been simplified.

- (Contrib) `exporter/loadbalancing`: Change AWS Cloud map resolver config fields from camelCase to snake_case. ([#32331](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/32331))
  The snake_case is required in OTel Collector config fields. It used to be enforced by tests in cmd/oteltestbedcol,
  but we had to disable them. Now, the tests are going to be enforced on every component independently.
  Hence, the camelCase config fields recently added with the new AWS Cloud Map resolver has to be fixed.

### 💡 Enhancements 💡

- (Splunk) Include [`splunk-otel-dotnet`](https://github.com/signalfx/splunk-otel-dotnet) in the `splunk-otel-auto-instrumentation` deb/rpm packages (x86_64/amd64 only) ([#4679](https://github.com/signalfx/splunk-otel-collector/pull/4679))
  - **Note**: Only manual activation/configuration for .NET auto instrumentation is currently supported. See [README.md](https://github.com/signalfx/splunk-otel-collector/blob/main/instrumentation/README.md) for details.
- (Splunk) Update splunk-otel-javaagent to `v1.32.0` ([#4715](https://github.com/signalfx/splunk-otel-collector/pull/4715))

> - (Splunk) Enable collecting MSI information on Windows in the support bundle ([#4710](https://github.com/signalfx/splunk-otel-collector/pull/4710))

Author: I'm not sure if this should be in the changelog since it's support targeted, but I added it just in case. If we don't want it let me know.

  The modules will work the same as other Collector components.

- (Core) `mdatagen`: enable goleak tests by default via mdatagen ([#9959](https://github.com/open-telemetry/opentelemetry-collector/pull/9959))
- (Core) `cmd/mdatagen`: support excluding some metrics based on string and regexes in resource_attributes ([#9661](https://github.com/open-telemetry/opentelemetry-collector/pull/9661))
- (Core) `cmd/mdatagen`: Generate config and factory tests covering their requirements. ([#9940](https://github.com/open-telemetry/opentelemetry-collector/pull/9940))
  The tests are moved from cmd/builder.
- (Core) `confmap`: Add `ProviderSettings`, `ConverterSettings`, `ProviderFactories`, and `ConverterFactories` fields to `confmap.ResolverSettings` ([#9516](https://github.com/open-telemetry/opentelemetry-collector/pull/9516))
  This allows configuring providers and converters, which are instantiated by `NewResolver` using the given factories.
- (Contrib) `vcenterreceiver`: Changes process for collecting VMs & VM perf metrics used by the `vccenterreceiver` to be more efficient (one call now for all VMs) ([#31837](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31837))

> - (Contrib) `splunkhecreceiver`: adding support for ack in the splunkhecreceiver ([#26376](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/26376))

Author: Not sure if we should include this in our changelog since we don't have the ack extension in our distribution. We should probably be adding the ack extension.

Reviewer 2: Adding the ack extension given that splunkhec can now work with it makes sense to me 👍

Author: PR posted: https://github.com/signalfx/splunk-otel-collector/pull/4724

  This change was made because it greatly reduces the CPU usage of the process and processes scrapers.
- (Contrib) `filelogreceiver`: Add `send_quiet` and `drop_quiet` options for `on_error` setting of operators ([#32145](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/32145))
- (Contrib) `pkg/ottl`: Add `IsList` OTTL Function ([#27870](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/27870))
- (Contrib) `filelogreceiver`: Add `exclude_older_than` configuration setting ([#31053](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31053))
- (Contrib) `pkg/stanza/operator/transformer/recombine`: add a new "max_unmatched_batch_size" config parameter to configure the maximum number of consecutive entries that will be combined into a single entry before the match occurs ([#31653](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31653))

### 🧰 Bug fixes 🧰

> - (Splunk) `receiver/discovery`: Fix locking mechanism on attributes ([#4712](https://github.com/signalfx/splunk-otel-collector/pull/4712))

Author: ```suggestion
- (Splunk) `receiver/discovery`: Fix locking mechanism on attributes ([#4712](https://github.com/signalfx/splunk-otel-collector/pull/4712))
- (Splunk) MSI elevated install failing: installs that required elevation were fixed. ([#4688](https://github.com/signalfx/splunk-otel-collector/pull/4688))
```

- (Splunk) Include [`splunk-otel-dotnet`](https://github.com/signalfx/splunk-otel-dotnet) in the `splunk-otel-auto-instrumentation` deb/rpm packages (x86_64/amd64 only) ([#4679](https://github.com/signalfx/splunk-otel-collector/pull/4679))
  - **Note**: Only manual activation/configuration for .NET auto instrumentation is currently supported. See [README.md](https://github.com/signalfx/splunk-otel-collector/blob/main/instrumentation/README.md) for details.
- (Splunk) Update splunk-otel-javaagent to `v1.32.0` ([#4715](https://github.com/signalfx/splunk-otel-collector/pull/4715))
- (Splunk) Enable collecting MSI information on Windows in the support bundle ([#4710](https://github.com/signalfx/splunk-otel-collector/pull/4710))
- (Splunk) Bump version of bundled Python to 3.11.9 ([#4729](https://github.com/signalfx/splunk-otel-collector/pull/4729))

> - (Splunk) `receiver/mongodb`: Enable auto-discovery ([#4722](https://github.com/signalfx/splunk-otel-collector/pull/4722))

Author: ```suggestion
- (Splunk) `receiver/mongodb`: Enable auto-discovery when TLS is disabled ([#4722](https://github.com/signalfx/splunk-otel-collector/pull/4722))
```

- (Splunk) Include [`splunk-otel-dotnet`](https://github.com/signalfx/splunk-otel-dotnet) in the `splunk-otel-auto-instrumentation` deb/rpm packages (x86_64/amd64 only) ([#4679](https://github.com/signalfx/splunk-otel-collector/pull/4679))
  - **Note**: Only manual activation/configuration for .NET auto instrumentation is currently supported. See [README.md](https://github.com/signalfx/splunk-otel-collector/blob/main/instrumentation/README.md) for details.
- (Splunk) Update splunk-otel-javaagent to `v1.32.0` ([#4715](https://github.com/signalfx/splunk-otel-collector/pull/4715))

> - (Splunk) Enable collecting MSI information on Windows in the support bundle ([#4710](https://github.com/signalfx/splunk-otel-collector/pull/4710))

Author: ```suggestion
- (Splunk) Enable collecting MSI information on Windows in the support bundle ([#4710](https://github.com/signalfx/splunk-otel-collector/pull/4710))
- (Splunk) Bump version of bundled Python to 3.11.9 ([#4729](https://github.com/signalfx/splunk-otel-collector/pull/4729))
```

- (Splunk) Include [`splunk-otel-dotnet`](https://github.com/signalfx/splunk-otel-dotnet) in the `splunk-otel-auto-instrumentation` deb/rpm packages (x86_64/amd64 only) ([#4679](https://github.com/signalfx/splunk-otel-collector/pull/4679))
  - **Note**: Only manual activation/configuration for .NET auto instrumentation is currently supported. See [README.md](https://github.com/signalfx/splunk-otel-collector/blob/main/instrumentation/README.md) for details.
- (Splunk) Update splunk-otel-javaagent to `v1.32.0` ([#4715](https://github.com/signalfx/splunk-otel-collector/pull/4715))
- (Splunk) Enable collecting MSI information on Windows in the support bundle ([#4710](https://github.com/signalfx/splunk-otel-collector/pull/4710))

> - (Splunk) Bump version of bundled Python to 3.11.9 ([#4729](https://github.com/signalfx/splunk-otel-collector/pull/4729))

Author: ```suggestion
- (Splunk) Bump version of bundled Python to 3.11.9 ([#4729](https://github.com/signalfx/splunk-otel-collector/pull/4729))
- (Splunk) `receiver/mongodb`: Enable auto-discovery ([#4722](https://github.com/signalfx/splunk-otel-collector/pull/4722))
```

- (Splunk) Update splunk-otel-javaagent to `v1.32.0` ([#4715](https://github.com/signalfx/splunk-otel-collector/pull/4715))
- (Splunk) Enable collecting MSI information on Windows in the support bundle ([#4710](https://github.com/signalfx/splunk-otel-collector/pull/4710))
- (Splunk) Bump version of bundled Python to 3.11.9 ([#4729](https://github.com/signalfx/splunk-otel-collector/pull/4729))
- (Splunk) `receiver/mongodb`: Enable auto-discovery ([#4722](https://github.com/signalfx/splunk-otel-collector/pull/4722))
- (Core) `confighttp`: Disable concurrency in zstd compression ([#8216](https://github.com/open-telemetry/opentelemetry-collector/pull/8216))
- (Core) `cmd/builder`: Allow configuring `confmap.Provider`s in the builder. ([#4759](https://github.com/open-telemetry/opentelemetry-collector/pull/4759))
  If no providers are specified, the defaults are used.
  The default providers are: env, file, http, https, and yaml.

  To configure providers, use the `providers` key in your OCB build
  manifest with a list of Go modules for your providers.

>   The modules will work the same as other Collector components.

Reviewer: Same comment. We should not list the builder changes

> - (Core) `mdatagen`: enable goleak tests by default via mdatagen ([#9959](https://github.com/open-telemetry/opentelemetry-collector/pull/9959))

Reviewer: Not an end-user facing change

- (Core) `cmd/mdatagen`: support excluding some metrics based on string and regexes in resource_attributes ([#9661](https://github.com/open-telemetry/opentelemetry-collector/pull/9661))
- (Core) `cmd/mdatagen`: Generate config and factory tests covering their requirements. ([#9940](https://github.com/open-telemetry/opentelemetry-collector/pull/9940))

>   The tests are moved from cmd/builder.

Reviewer: Should not be included. We don't ship mdatagen binary itself

- (Core) `confmap`: Add `ProviderSettings`, `ConverterSettings`, `ProviderFactories`, and `ConverterFactories` fields to `confmap.ResolverSettings` ([#9516](https://github.com/open-telemetry/opentelemetry-collector/pull/9516))

>   This allows configuring providers and converters, which are instantiated by `NewResolver` using the given factories.

Reviewer: Go API change. I'd rather exclude this
